lib/inventory/internal/delay/multi.go

```diff
@@ -0,0 +1,210 @@
+	VariableInterval *interval.VariableDuration
+	// FirstJitter is the jitter applied to the first interval. It's not applied
+	// to the interval after the first tick. If unset, the standard jitter is
+	// applied to the first interval.
+	FirstJitter retryutils.Jitter
+	// Jitter is a jitter function, applied every tick (if set) to the fixed or
+	// variable interval (except for the first tick if FirstJitter is set).
+	Jitter retryutils.Jitter
+
+	clock clockwork.Clock
+}
+
+// Multi is a ticker-like abstraction around a [*time.Timer] that's made to tick
+// periodically with a potentially varying interval and optionally some jitter.
+// Its use requires some care as the logic driving the ticks and the jittering
+// must be explicitly invoked by the code making use of it, but uses no
+// background resources. It tracks an arbitrary number of sub-intervals by key,
+// allowing a single delay to be applied to multiple overlapping intervals.
+type Multi[T comparable] struct {
+	clock clockwork.Clock
+	timer clockwork.Timer
+
+	heap heap[entry[T]]
+
+	fixedInterval    time.Duration
+	variableInterval *interval.VariableDuration
+
+	firstJitter retryutils.Jitter
+	jitter      retryutils.Jitter
+}
+
+// NewMulti returns a new [*Multi]. Note that the delay starts with no subintervals
+// and will not tick until at least one subinterval is added.
+func NewMulti[T comparable](p MultiParams) *Multi[T] {
+	if p.clock == nil {
+		p.clock = clockwork.NewRealClock()
+	}
+	return &Multi[T]{
+		clock: p.clock,
+
+		heap: heap[entry[T]]{
+			Less: entryLess[T],
+		},
+
+		fixedInterval:    p.FixedInterval,
+		variableInterval: p.VariableInterval,
+
+		firstJitter: p.FirstJitter,
+		jitter:      p.Jitter,
+	}
+}
+
+func (h *Multi[T]) Add(key T) {
+	// add new target to the heap
+	now := h.clock.Now()
+	entry := entry[T]{
+		tick: now.Add(h.interval(true /* first */)),
+		key:  key,
+	}
+	h.heap.Push(entry)
+
+	// trigger reset in case the new entry should be the next target
+	h.reset(now, false /* fired */)
+}
+
+func (h *Multi[T]) Remove(key T) {
+	// key is not the current target, remove it from the heap
+	for i, entry := range h.heap.Slice {
+		if entry.key == key {
+			h.heap.Remove(i)
+			if i == 0 {
+				// if the removed entry was the next target, we need to reset
+				// the underlying timer to the new target.
+				h.reset(h.clock.Now(), false /* fired */)
+			}
+			return
+		}
+	}
+}
+
+// Tick *must* be called exactly once for each firing observed on the Elapsed channel, with the time
+// of the firing. Tick will advance the internal state of the multi to start targeting the next interval,
+// and return the key associated with the interval that just fired.
+func (h *Multi[T]) Tick(now time.Time) (key T) {
+	// advance the current root entry (source of the tick), and record its
+	// key for later return.
+	root := h.heap.Root()
+	key = root.key
+	root.tick = now.Add(h.interval(false /* first */))
```

Comment:
Should we bump the tick up from itself rather than from `now`?
```suggestion
	root.tick = root.tick.Add(h.interval(false /* first */))
```

Reply:
I kind of like it this way.  If perf is good, the difference is inconsequential and we might as well calculate from the variable that isn't behind a pointer.  If perf is bad, this adds a bit of slack by always scheduling the N+1th firing a full duration after then Nth firing, even if the Nth firing was delayed in being observed for some reason.  For the intended usecase, I think thats a desirable property, and I think its as sensible a behavior as anything else given that the API doesn't try to hide the fact that we're not resetting until `Tick` is called.

Reply:
I don't know if I agree - this is not a goroutine that's only tightly looping on the heartbeats, we might be stuck somewhere else for a bit before coming back to the timer, couldn't we?

If choosing the correct time is inconsequential, why are we bringing over the timestamp from the timer instead of just calling Now()?